juice/src/net/config.rs

```diff
@@ -0,0 +1,14 @@
+/// A configuration of the layer.
+/// Determines the type of the layer and the (optional) layer settings.
+#[derive(Debug, Clone)]
+pub enum LayerConfig {
```

Comment:
Yes, I saw the conversation in #125. Right now the net creation is done by the user indirectly by creating the config first and then letting the framework to build requested layers, connecting them in the process. If we want to support external layers within the same paradigm, we'll need:
1. A way to represent a config of the external layer. Something like:
```rust
pub trait UserLayerConfig {}

pub enum LayerConfig {
  Relu,
  UserLayer(Box<UserLayerConfig>),
}
``` 
2. A factory that can build user layers:
```rust
pub trait UserLayerFactory {
  fn create_layer(config: &dyn UserLayerConfig) -> Box<Layer>;
}
```
3. Which then can be passed to `Network::from_config()`:
```rust
pub fn layer_from_config<B: IBackend + LayerOps<f32> + 'static>(
    descriptor: Descriptor,
    config: &LayerConfig,
    user_layer_factory: &dyn UserLayerFactory
) -> Box<dyn Layer<B>> { ... } 
```

All of that looks a bit Frankenstein-ish. 

Maybe the right way is to go all in on layer factories:
1. We store layer config as `Any` objects.
2. Framework maintains a collection of layer factories, one of them being "system" and others can be registered by users.
3. When building network from config, each factory is tried one by one until one of them recognizes the config and creates the layer.
```rust
pub struct NetworkConfig {
  pub top: Box<Any>,
}

pub trait LayerFactory {
  // If this factory recognizes the passed config, it will create a layer. Otherwise returns None.
  fn try_create_layer(config: &dyn Any) -> Option<Box<Layer>>;
}

pub fn register_layer_factory(factory: Box<dyn LayerFactory>);

pub struct SystemLayerFactory {}

impl LayerFactory for SystemLayerFactory {
  fn try_create_layer(config: &dyn Any) -> Option<Box<Layer>> {
    if config.type_id() == TypeId::of::<ReluConfig>() {
      // Create ReLU layer.
    }
    ...
}
```

Yet another way is somehow make network constructions direct, i.e., instead of creating layer configs users will create layers directly. 

I'll need to ponder this a little more.

Reply:
I have to admit the added complexity is not appealing at this point, so I'd postpone the discussion for now. Generally I'd like to avoid any use of `Any` as much as possible.

Reply:
Creating layers directly would be an option, but this requires backend to be only passed in at the actual inference/backprop calls rather than the layer holding a reference, but usually there is some state related to a layer which requires backend specific memory, which would further complicate things. On the other hand, it's desirable to have the network declaration independent of the execution to some extent, but I am not sure how relevant this is practically. The main use case would ve to be able to declare a network without having a backend instance.

Reply:
How about this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=72938fbabd9d9fd6333bf086282d4504. Best to start reading from `main()` as it is driving some of the functions syntax.

Main differences:
1. No layer configs -- layers are now created directly; whatever configuration they need is passed to their constructors.
2. Backend is now passed as `dyn` ref, so layers do not depend on a specific backend.

Open questions:
1. How to deal with backend plugins. Obviously we can't have the same flexibility when each layer could have requested its own set of plugin functions. OTOH, this flexibility was nixed by the fact the Sequential had to request a set of plugin features which was a union of all possible children (`LayerOps`). So I think we can follow the same path and just use `LayerOps` here without losing anything.
2. Serialization. Essentially this requires the serialization engine to support deserializing trait objects. Need to check if capnp supports it.